core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java

```diff
@@ -333,8 +333,7 @@ private void publishDownstream(T item, boolean head) {
                 subscribed = true;
                 publisher.subscribe(this, executor, options);
             } else {
-                assert upstreamRequested > 0;
-                if (upstreamRequested < Long.MAX_VALUE) {
```

Comment:
```java
class ChannelSendOperator {

		@Override
		public void subscribe(Subscriber<? super T> writeSubscriber) {
			synchronized (this) {
				Assert.state(this.writeSubscriber == null, "Only one write subscriber supported");
				this.writeSubscriber = writeSubscriber;
				if (this.error != null || this.completed) {
					this.writeSubscriber.onSubscribe(Operators.emptySubscription());
					emitCachedSignals(); // 👈👈 
				}

		private boolean emitCachedSignals() {
			T item = this.item;
			this.item = null;
			if (item != null) {
				requiredWriteSubscriber().onNext(item); // 👈👈 
			}
}
```
@ikhoon , I found that spring's ChannelSendOperator calls downstream.onNext(item) without any request on below case!

1. ChannelSendOperator.`subscribe()`
2. ChannelSendOperator is already `completed`
3. emitCachedSignals()
4. downstream.onNext(item) // although there's no request, onNext() is invoked!

"ChannelSendOperator(upstream) can downstream.onNext(item) without any request" -> it sounds weird, but we can't change spring ChannelSendOperator's behavior so I just removed `assert upstreamRequested > 0;` on L336 ~! PTAL 🙇

Reply:
I see, changes still look good to me 👍

Reply:
> "ChannelSendOperator(upstream) can downstream.onNext(item) without any request"

It might be not related to this PR but the assert found the bug. Let me add some logs and debug to know why `ChannelSendOperator` was completed before `subscribe()`

Reply:
@ikhoon nim gentle ping 🙇

Can you check debug logs on broken test? I think I still can't check debug logs on CI/CD result.
I want to merge this PR and handle next PRs!

Reply:
You need to download artifacts to see debug logs on the summary page of the GitHub Actions job.
Otherwise, let me try to merge #5104 quickly to easily see the test logs.

Reply:
@injae-kim You can now check out the test results on Gradle build scans.
https://github.com/line/armeria/actions/runs/5901062192
https://ge.armeria.dev/s/r2qfke4r6ueky/tests/task/:spring:boot2-webflux-autoconfigure:shadedTest/details/com.linecorp.armeria.spring.web.reactive.MatrixVariablesTest/foo()?top-execution=1

Reply:
> test results on Gradle build scans.

oh it's cool! thanks a lot 🙇 I'll check test result and share to you~!!

Reply:
I suspected the problem is a bug in `ChannelSendOperator` that violates the most important rule 1.1. https://github.com/reactive-streams/reactive-streams-jvm/blob/v1.0.4/README.md#1.1
So I forked `ChannelSendOperator` and modified not to publish the cached item before a request is made. https://github.com/line/armeria/pull/4727/files#r1299525470

Reply:
The problem occurs only in Spring Boot 2 WebFlux. The following command can reproduce it locally.
```
./gradlew :spring:boot3-webflux-autoconfigure:testClasses :spring:boot2-webflux-autoconfigure:test --tests "*.MatrixVariablesTest"
```
`ChannelSendOperator` implementation has not changed between Spring Boot 2 and Spring Boot 3. The main difference was https://github.com/spring-projects/spring-framework/issues/28398 which avoids collecting `Flux` before passing to `ChannelSendOperator`. Consequently, `WriteBarrier.onComplete()` was not invoked before `WriteBarrier.subscribe()` is called and backpressure looks respected.

Reply:
Really thanks for your detailed investigation 🙇 If there's something I have to do, please tell me!